### PR TITLE
Add support for app.config.js in jest-expo

### PR DESCRIPTION
**Pull request by EvanBacon**
# Why

- Creates possible resolution for https://github.com/expo/expo/issues/4156
- Moved from https://github.com/expo/expo/pull/7447

# How

Use `getConfig` from `@expo/config` to read `app.config.js`.

# Test Plan

Tested in a different project.

### CODE REVIEW

File: packages/jest-expo/src/preset/createMockConstants.js [hunk 1]
```diff
@@ -11,15 +10,14 @@ const path = require('path');
  */
 module.exports = function createMockConstants() {
   const appConfig = _readAppConfiguration();
-  const expoConfig = (appConfig && appConfig.expo) || {};
+  const expoConfig = appConfig || {};
```
Comment (ide) [inline packages/jest-expo/src/preset/createMockConstants.js:13]:
```suggestion
  const expoConfig = _readAppConfiguration() ?? {};
```

File: packages/jest-expo/src/preset/createMockConstants.js [hunk 2]
```diff
@@ -35,7 +33,7 @@ module.exports = function createMockConstants() {
 };
 
 function _readAppConfiguration() {
```
Comment (ide) [inline packages/jest-expo/src/preset/createMockConstants.js:35]:
```suggestion
function _readExpoConfiguration() {
```

File: packages/jest-expo/src/preset/createMockConstants.js [hunk 3]
```diff
@@ -44,22 +42,20 @@ function _readAppConfiguration() {
   do {
     currentDirectory = nextDirectory;
 
-    const candidatePath = path.join(currentDirectory, 'app.json');
-    json = _safeReadFile(candidatePath, 'utf8');
+    try {
+      config = getConfig(currentDirectory);
+    } catch (e) {
+      if (!e.message.includes('expected package.json path')) {
```
Comment (ide) [inline packages/jest-expo/src/preset/createMockConstants.js:48]: This seems really fragile. Someone could change the config package or whatever throws this error and break this code that's in a different repo. Is there some way we could avoid getting this error up front, or at least make it part of the public API (e.g. with an error code) of getConfig?

**Review by ide — CHANGES_REQUESTED** [review]

**Comment by pdugan20** [comment]
Is there a workaround for this or any idea as to why it might be merged?

**Comment by illBeRoy** [comment]
ping! this seems like a really good idea...

**Comment by mvanroon** [comment]
ping @EvanBacon @byCedric

**Comment by brentvatne** [comment]
if someone wants to update this pr and reopen that'd be great, we're busy with some other projects right now

**Review by brentvatne — APPROVED** [review]

**Pull request merged**